### PR TITLE
Add loading indicator and counter to query result

### PR DESCRIPTION
**Pull request by abbyhu2000**
### Description

* when a query takes longer than 3s to complete, a loading indicator will appear after 3s and will also show a time counter; the completion time will show in seconds.
* when a query takes less than 1s to complete, the completion time will be shown in miliseconds. 

## Screenshot

https://github.com/user-attachments/assets/60e9d207-7cd4-4e14-b0f6-42a5a6e6c993

## Changelog

- feat: Add loading indicator and counter to query result

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx
Comment (ashwin-pc) [inline src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:1]: Can you add a unit test for this functionality?

**Review by ashwin-pc — COMMENTED** [review]
Nice! I like the how you have accounted for the different ways the user will use and interact with the UI here. Just one omment on tests, the rest of the PR looks good

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx [hunk 1]
```diff
@@ -28,21 +28,60 @@ export interface QueryStatus {
...
+    } else {
+      setElapsedTime(0);
+    }
+  };
+
+  useEffect(() => {
+    const interval = setInterval(getTime, 1000);
+
+    return () => clearInterval(interval);
+  });
```
Comment (sejli) [inline src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:57]: Does this need to run on every render?

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx
Comment (abbyhu2000) [reply src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:1]: added

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx [hunk 2]
```diff
@@ -28,21 +28,66 @@ export interface QueryStatus {
...
+
 export function QueryResult(props: { queryStatus: QueryStatus }) {
   const [isPopoverOpen, setPopover] = useState(false);
+  const [elapsedTime, setElapsedTime] = useState(0);
   const onButtonClick = () => {
     setPopover(!isPopoverOpen);
   };
 
+  const getTime = () => {
+    const time = new Date().getTime() - props.queryStatus.startTime!;
```
Comment (AMoo-Miki) [inline src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:45]:
```suggestion
    const time = Date.now() - props.queryStatus.startTime!;
```

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx [hunk 3]
```diff
@@ -28,21 +28,66 @@ export interface QueryStatus {
...
+    if (time > BUFFER_TIME) {
+      setElapsedTime(time);
```
Comment (AMoo-Miki) [inline src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:47]: Based on the interval below and the logic here, elapsedTime will be
0 for first run
0 for second run
3000 for third run
4000 for fourth run

Is that expected?

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx [hunk 4]
```diff
@@ -28,21 +28,66 @@ export interface QueryStatus {
     statusCode?: number;
   };
   elapsedMs?: number;
+  startTime?: number;
 }
 
+// This is the time in milliseconds that the query will wait before showing the loading spinner
```
Comment (AMoo-Miki) [inline src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:34]: Why would it wait for this huge amount of time tow show the spinner if there really is something loading?

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx [hunk 5]
```diff
@@ -28,21 +28,66 @@ export interface QueryStatus {
...
+  }
+
   if (props.queryStatus.status === ResultStatus.READY) {
+    let message;
+    if (!props.queryStatus.elapsedMs) {
+      message = 'Completed';
+    } else if (props.queryStatus.elapsedMs < 1000) {
+      message = `Completed in ${props.queryStatus.elapsedMs} ms`;
+    } else {
+      message = `Completed in ${Math.floor(props.queryStatus.elapsedMs / 1000)} s`;
```
Comment (AMoo-Miki) [inline src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:84]: Not sure what the UX is but as a user interested it knowing how long it took, flooring to the second is not good enough.

```suggestion
      message = `Completed in ${(props.queryStatus.elapsedMs / 1000).toFixed(1)} s`;
```

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx [hunk 6]
```diff
@@ -28,21 +28,66 @@ export interface QueryStatus {
...
+      >
+        {`Loading ${time} s`}
+      </EuiButtonEmpty>
+    );
+  }
+
   if (props.queryStatus.status === ResultStatus.READY) {
+    let message;
+    if (!props.queryStatus.elapsedMs) {
+      message = 'Completed';
```
Comment (AMoo-Miki) [inline src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:80]: i18n please for all of these `message` values.

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx [hunk 7]
```diff
@@ -55,7 +100,13 @@ export function QueryResult(props: { queryStatus: QueryStatus }) {
...
-        <EuiButtonEmpty iconSide="left" iconType={'alert'} size="xs" onClick={onButtonClick}>
+        <EuiButtonEmpty
+          iconSide="left"
+          iconType={'alert'}
+          size="xs"
+          onClick={onButtonClick}
+          data-test-subj="queryResultErrorBtn"
+        >
           <EuiText size="xs" color="subdued">
             {'Error'}
```
Comment (AMoo-Miki) [inline src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:111]: i18n plz

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx [hunk 8]
```diff
@@ -65,6 +116,7 @@ export function QueryResult(props: { queryStatus: QueryStatus }) {
       closePopover={() => setPopover(false)}
       panelPaddingSize="s"
       anchorPosition={'downRight'}
+      data-test-subj="queryResultError"
     >
       <EuiPopoverTitle>ERRORS</EuiPopoverTitle>
```
Comment (AMoo-Miki) [inline src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:121]: i18n plz

File: src/plugins/discover/public/application/view_components/utils/use_search.ts [hunk 9]
```diff
@@ -119,12 +120,14 @@ export const useSearch = (services: DiscoverViewServices) => {
     );
   }, [savedSearch, services.uiSettings, timefilter]);
 
+  const startTime = new Date().getTime();
```
Comment (AMoo-Miki) [inline src/plugins/discover/public/application/view_components/utils/use_search.ts:123]:
```suggestion
  const startTime = Date.now();
```

File: src/plugins/discover/public/application/view_components/utils/use_search.ts [hunk 10]
```diff
@@ -161,7 +164,6 @@ export const useSearch = (services: DiscoverViewServices) => {
     dataset = searchSource.getField('index');
 
     let elapsedMs;
-
     try {
       // Only show loading indicator if we are fetching when the rows are empty
       if (fetchStateRef.current.rows?.length === 0) {
```
Comment (AMoo-Miki) [inline src/plugins/discover/public/application/view_components/utils/use_search.ts:169]: if `fetchStateRef.current` has no rows, `undefined === 0` will be false. I think what you really want is
```suggestion
      if (!(fetchStateRef.current.rows?.length)) {
```

File: yarn.lock [hunk 11]
```diff
@@ -15981,7 +15981,7 @@ string-similarity@^4.0.1:
   resolved "https://registry.yarnpkg.com/string-similarity/-/string-similarity-4.0.4.tgz#42d01ab0b34660ea8a018da8f56a3309bb8b2a5b"
   integrity sha512-/q/8Q4Bl4ZKAPjj8WerIBJWALKkaPRfrvhfF8k/B23i4nzrlRj2/go1m90In7nG/3XDSbOo0+pu6RvCTM9RGMQ==
 
-"string-width-cjs@npm:string-width@^4.2.0":
+"string-width-cjs@npm:string-width@^4.2.0", string-width@^4.1.0, string-width@^4.2.0, string-width@^4.2.3:
```
Comment (AMoo-Miki) [inline yarn.lock:15984]: plz undo these unrelated changes.

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx (hunk 4 above)
Comment (abbyhu2000) [reply src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:34]: @ashwin-pc requests that we wait for some time before we show the loading indicator.

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx [hunk 12]
```diff
@@ -28,21 +28,66 @@ export interface QueryStatus {
...
+const BUFFER_TIME = 3000;
+
 export function QueryResult(props: { queryStatus: QueryStatus }) {
   const [isPopoverOpen, setPopover] = useState(false);
+  const [elapsedTime, setElapsedTime] = useState(0);
   const onButtonClick = () => {
     setPopover(!isPopoverOpen);
   };
 
+  const getTime = () => {
```
[the remark this answers is not in the document]
Comment (abbyhu2000) [reply src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:44]: updated

File: yarn.lock (hunk 11 above)
Comment (abbyhu2000) [reply yarn.lock:15984]: removed

File: src/plugins/discover/public/application/view_components/utils/use_search.ts (hunk 10 above)
Comment (abbyhu2000) [reply src/plugins/discover/public/application/view_components/utils/use_search.ts:169]: this PR did not change this line. 

but i think we want the if statement here only check the case: `length === 0`

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx (hunk 3 above)
Comment (abbyhu2000) [reply src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:47]: i think 1s interval is a reasonable time to get elapsedTime here? since @ashwin-pc also requests we show loading time in seconds

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx (hunk 4 above)
Comment (ashwin-pc) [reply src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:34]: Yeah for most usecases the query runs in milliseconds, its only when the query takes longer than a few seconds that you are left wondering whats happening. I think 3 seconds is a good buffer to wait before indicating the user that hey your query is running, its just taking a bit longer than expected.

File: src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx (hunk 3 above)
Comment (ashwin-pc) [reply src/plugins/data/public/query/query_string/language_service/lib/query_result.tsx:47]: Its rounded down to seconds so that should not be a problem. Take a look at the video attached Miki :)

**Review by ashwin-pc — APPROVED** [review]
Nice!

**Review by LDrago27 — APPROVED** [review]

**Comment by abbyhu2000** [comment]
Link checker failure not related: 
```
* [404] [http://sites.google.com/site/murmurhash/](http://sites.google.com/site/murmurhash/) | Failed: Network error: Not Found
